test/e2e/framework/network_pods.go

```diff
@@ -1,18 +1,150 @@
 package framework
 
 import (
+	"fmt"
 	"strconv"
+	"time"
 
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	. "github.com/onsi/gomega"
 )
 
+type NetworkPodType int
+
+const (
+	InvalidPodType NetworkPodType = iota
+	ListenerPod
+	ConnectorPod
+)
+
+type NetworkPodScheduling int
+
+const (
+	InvalidScheduling NetworkPodScheduling = iota
+	GatewayNode
+	NonGatewayNode
+)
+
+type NetworkPodConfig struct {
+	Type       NetworkPodType
+	Cluster    ClusterIndex
+	Scheduling NetworkPodScheduling
+	Port       int
+	Data       string
+	RemoteIP   string
+	// TODO: namespace, once https://github.com/submariner-io/submariner/pull/141 is merged
+}
+
+type NetworkPod struct {
+	Pod                *v1.Pod
+	Config             *NetworkPodConfig
+	TerminationCode    int32
+	TerminationMessage string
+	framework          *Framework
+}
+
+const (
+	TestPort = 1234
+)
+
+func (f *Framework) NewNetworkPod(config *NetworkPodConfig) *NetworkPod {
+
+	// check if all necessary details are provided
+	Expect(config.Scheduling).ShouldNot(Equal(InvalidScheduling))
+	Expect(config.Type).ShouldNot(Equal(InvalidPodType))
+
+	// setup unset defaults
+	if config.Port == 0 {
+		config.Port = TestPort
+	}
+
+	if config.Data == "" {
+		config.Data = string(uuid.NewUUID())
+	}
+
+	networkPod := &NetworkPod{Config: config, framework: f, TerminationCode: -1}
+
+	switch config.Type {
+	case ListenerPod:
+		networkPod.buildTCPCheckListenerPod()
+	case ConnectorPod:
+		networkPod.buildTCPCheckConnectorPod()
+	}
+
+	return networkPod
+}
+
+func (np *NetworkPod) AwaitReady() {
+	var finalPod *v1.Pod
+	pc := np.framework.ClusterClients[np.Config.Cluster].CoreV1().Pods(np.framework.Namespace)
+	err := wait.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
+		pod, err := pc.Get(np.Pod.Name, metav1.GetOptions{})
+		if err != nil {
+			if IsTransientError(err) {
+				Logf("Transient failure when attempting to list pods: %v", err)
+				return false, nil // return nil to avoid PollImmediate from stopping
+			}
+			return false, err
+		}
+		if pod.Status.Phase != v1.PodRunning {
+			if pod.Status.Phase != v1.PodPending {
+				return false, fmt.Errorf("expected pod to be in phase \"Pending\" or \"Running\"")
+			}
+			return false, nil // pod is still pending
+		}
+		finalPod = pod
+		return true, nil // pods is running
+	})
+	Expect(err).NotTo(HaveOccurred())
+	np.Pod = finalPod
+}
+
+func (np *NetworkPod) AwaitSuccessfulFinish() {
```

Comment:
My version was more generic though. For example being able to wait for non-successful makes it possible to do negative tests in network policies (pods don't connect), but we can add another function on a later PR very easily. The resulting test code is cleaner your way.

Reply:
yeah that's what I was thinking, ie add `AwaitFinish` later if needed.